src/BuildLoop.cpp

```diff
@@ -150,27 +160,237 @@ static void SignalMainThreadToStartCleaningUp(BuildQueue *queue)
     MutexUnlock(&queue->m_BuildFinishedMutex);
 }
 
-static void AdvanceNode(BuildQueue *queue, ThreadState *thread_state, RuntimeNode *node, Mutex *queue_lock)
+static void FinishNode(BuildQueue* queue, RuntimeNode* node)
+{
+    CheckHasLock(&queue->m_Lock);
+
+    node->m_Finished = true;
+    queue->m_FinishedNodeCount++;
+
+    if (RuntimeNodeIsExplicitlyRequested(node))
+        queue->m_FinishedRequestedNodeCount++;
+
+    RuntimeNodeFlagInactive(node);
+    if (queue->m_FinishedRequestedNodeCount == queue->m_Config.m_AmountOfRuntimeNodesSpecificallyRequested)
+        SignalMainThreadToStartCleaningUp(queue);
+
+    EnqueueDependeesWhoMightNowHaveBecomeReadyToRun(queue, node);
+}
+
+static bool IsNodeCacheableByLeafInputsAndCachingEnabled(BuildQueue* queue, RuntimeNode* node)
+{
+    if (!queue->m_Config.m_AttemptCacheReads && !queue->m_Config.m_AttemptCacheWrites)
+        return false;
+    return 0 != (node->m_DagNode->m_Flags & Frozen::DagNode::kFlagCacheableByLeafInputs);
+}
+
+static void AttemptCacheWrite(BuildQueue* queue, ThreadState* thread_state, RuntimeNode* node)
+{
+    CheckDoesNotHaveLock(&queue->m_Lock);
+
+    uint64_t time_exec_started = TimerGet();
+
+    char path[kMaxPathLength];
+    char digestString[kDigestStringSize];
+    DigestToString(digestString, node->m_CurrentLeafInputSignature->digest);
+    snprintf(path, sizeof(path), "%s", digestString);
+    FILE *sig = fopen(path, "w");
+    if (sig == NULL)
+    {
+        printf("Failed to open file for signature ingredient writing. Skipping CacheWrite.\n");
+        return;
+    }
+
+    //we already calculated the leaf input signature before, but we'll do it again because now we want to have the ingredient stream written out to disk.
```

Comment:
Should this be a "TODO: reuse the previously calculated leaf input signature" comment?

Reply:
No, this is by design. We don't calculate the signature again, because we cannot fetch the old the signature, but because we previously generated it without generating the ingredient stream - which we did not know we would need in advance. We _could_ separate the generation of the ingredient stream and the signature, so we don't have to calculate the signature again, but doing so would give us messier code, with likely little gains in performance.